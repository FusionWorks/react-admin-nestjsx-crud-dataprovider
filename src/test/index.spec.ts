import createDataProvider, { encodeParamsInResource } from "../index";
import { asyncResultsToValue, createJestFnSerialReturn } from "./testsHelpers";

describe("Basic tests", () => {
  test("GET_ONE 1", async () => {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const httpClient: any = createJestFnSerialReturn([
      {
        json: {
          id: 42,
          title: "The Hitchhiker's guide to the galaxy"
        }
      }
    ]);

    const dataProvider = createDataProvider(
      "https://test.test/api",
      httpClient
    );

    expect(
      await dataProvider("GET_ONE", "books", { id: 42 })
    ).toMatchSnapshot();
    expect(await asyncResultsToValue(httpClient)).toMatchSnapshot();
  });
});

describe("CREATE 1", () => {
  test("CREATE 1", async () => {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const httpClient: any = createJestFnSerialReturn([
      {
        json: {
          id: 42,
          title: "The Hitchhiker's guide to the galaxy",
          pages: 3
        }
      },
      // Create gonna make another request to fetch the full record
      {
        json: {
          id: 42,
          title: "The Hitchhiker's guide to the galaxy",
          pages: 3,
          votes: []
        }
      }
    ]);

    const dataProvider = createDataProvider(
      "https://test.test/api",
      httpClient
    );

    expect(
      await dataProvider("CREATE", "books", {
        title: "The Hitchhiker's guide to the galaxy",
        pages: 3
      })
    ).toMatchSnapshot();
    expect(await asyncResultsToValue(httpClient)).toMatchSnapshot();
  });
});

describe("UPDATE 1", () => {
  test("UPDATE 1", async () => {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const httpClient: any = createJestFnSerialReturn([
      {
        json: {
          id: 42,
          title: "For Whom The Bell Tolls"
        }
      },
      {
        json: {
          id: 42,
          title: "For Whom The Bell Tolls",
          pages: [{ content: "blabla" }]
        }
      }
    ]);

    const dataProvider = createDataProvider(
      "https://test.test/api",
      httpClient
    );

    expect(
      await dataProvider("UPDATE", "books", {
        id: 42,
        data: { title: "For Whom The Bell Tolls" }
      })
    ).toMatchSnapshot();
    expect(await asyncResultsToValue(httpClient)).toMatchSnapshot();
  });
});

describe("UPDATE_MANY 1", () => {
  test("UPDATE_MANY 1", async () => {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const httpClient: any = createJestFnSerialReturn(
      [
        {
          id: 42,
          title: "The Hitchhiker's guide to the galaxy"
        },
        {
          id: 50,
          title: "The Hitchhiker's guide to the galaxy 2"
        },
        {
          id: 42,
          title: "The Hitchhiker's guide to the galaxy",
          pages: [{ content: "blabla" }]
        },
        {
          id: 50,
          title: "The Hitchhiker's guide to the galaxy 2",
          pages: [{ content: "blabla" }]
        }
      ].map(json => ({ json }))
    );

    const dataProvider = createDataProvider(
      "https://test.test/api",
      httpClient
    );

    expect(
      await dataProvider("UPDATE_MANY", "books", {
        ids: [42, 50],
        data: { views: 0 }
      })
    ).toMatchSnapshot();
    expect(await asyncResultsToValue(httpClient)).toMatchSnapshot();
  });
});

describe("DELETE_MANY 1", () => {
  test("DELETE_MANY 1", async () => {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const httpClient: any = createJestFnSerialReturn([{}, {}]);

    const dataProvider = createDataProvider(
      "https://test.test/api",
      httpClient
    );

    expect(
      await dataProvider("DELETE_MANY", "books", { ids: [42, 43] })
    ).toMatchSnapshot();
    expect(await asyncResultsToValue(httpClient)).toMatchSnapshot();
  });
});

describe("encodeParamsInResource tests", () => {
  test("encodeParamsInResource test 1", async () => {
    expect(
      encodeParamsInResource("books", {
        fields: ["id", "name", "year"],
        join: [
          {
            field: "pages",
            select: ["number", "words"]
          },
          {
            field: "author",
            select: ["id", "name"]
          },
          {
            field: "author.favoriteFood",
            select: ["id", "name"]
          }
        ]
      })
    ).toMatchSnapshot();
  });
});

describe("encoded params tests", () => {
  test("encoded params test 1", async () => {
    // @ts-ignore
    const httpClient: any = jest.fn(async (type, resource, params) => {
      return {
        json: {
          id: 1
        },
        ...params
      };
    });

    const dataProvider = createDataProvider(
      "https://test.test/api",
      httpClient
    );

    const resourceWithParams = encodeParamsInResource("books", {
      fields: ["id", "name", "year"],
      join: [
        {
          field: "pages",
          select: ["number", "words"]
        },
        {
          field: "author",
          select: ["id", "name"]
        },
        {
          field: "author.favoriteFood",
          select: ["id", "name"]
        }
      ]
    });

    expect(
      await dataProvider("GET_ONE", resourceWithParams, { id: "5" })
    ).toMatchSnapshot();
    expect(await asyncResultsToValue(httpClient)).toMatchSnapshot();
  });
});
