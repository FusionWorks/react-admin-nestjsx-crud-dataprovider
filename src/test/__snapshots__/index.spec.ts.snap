// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic tests GET_ONE 1 1`] = `
Object {
  "data": Object {
    "id": 42,
    "title": "The Hitchhiker's guide to the galaxy",
  },
}
`;

exports[`Basic tests GET_ONE 1 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://test.test/api/books/42?",
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 42,
          "title": "The Hitchhiker's guide to the galaxy",
        },
      },
    },
  ],
}
`;

exports[`CREATE 1 CREATE 1 1`] = `
Object {
  "data": Object {
    "id": 42,
    "pages": 3,
    "title": "The Hitchhiker's guide to the galaxy",
    "votes": Array [],
  },
}
`;

exports[`CREATE 1 CREATE 1 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://test.test/api/books",
      Object {
        "body": undefined,
        "method": "POST",
      },
    ],
    Array [
      "https://test.test/api/books/42?",
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 42,
          "pages": 3,
          "title": "The Hitchhiker's guide to the galaxy",
        },
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 42,
          "pages": 3,
          "title": "The Hitchhiker's guide to the galaxy",
          "votes": Array [],
        },
      },
    },
  ],
}
`;

exports[`DELETE_MANY 1 DELETE_MANY 1 1`] = `
Object {
  "data": Array [
    42,
    43,
  ],
}
`;

exports[`DELETE_MANY 1 DELETE_MANY 1 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://test.test/api/books/42",
      Object {
        "method": "DELETE",
      },
    ],
    Array [
      "https://test.test/api/books/43",
      Object {
        "method": "DELETE",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {},
    },
    Object {
      "type": "return",
      "value": Object {},
    },
  ],
}
`;

exports[`UPDATE 1 UPDATE 1 1`] = `
Object {
  "data": Object {
    "id": 42,
    "pages": Array [
      Object {
        "content": "blabla",
      },
    ],
    "title": "For Whom The Bell Tolls",
  },
}
`;

exports[`UPDATE 1 UPDATE 1 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://test.test/api/books/42",
      Object {
        "body": "{\\"title\\":\\"For Whom The Bell Tolls\\"}",
        "method": "PATCH",
      },
    ],
    Array [
      "https://test.test/api/books/42?",
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 42,
          "title": "For Whom The Bell Tolls",
        },
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 42,
          "pages": Array [
            Object {
              "content": "blabla",
            },
          ],
          "title": "For Whom The Bell Tolls",
        },
      },
    },
  ],
}
`;

exports[`UPDATE_MANY 1 UPDATE_MANY 1 1`] = `
Object {
  "data": Array [
    Object {
      "data": Object {
        "id": 42,
        "pages": Array [
          Object {
            "content": "blabla",
          },
        ],
        "title": "The Hitchhiker's guide to the galaxy",
      },
    },
    Object {
      "data": Object {
        "id": 50,
        "pages": Array [
          Object {
            "content": "blabla",
          },
        ],
        "title": "The Hitchhiker's guide to the galaxy 2",
      },
    },
  ],
}
`;

exports[`UPDATE_MANY 1 UPDATE_MANY 1 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://test.test/api/books/42",
      Object {
        "body": "{\\"views\\":0}",
        "method": "PUT",
      },
    ],
    Array [
      "https://test.test/api/books/50",
      Object {
        "body": "{\\"views\\":0}",
        "method": "PUT",
      },
    ],
    Array [
      "https://test.test/api/books/42?",
      Object {},
    ],
    Array [
      "https://test.test/api/books/50?",
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 42,
          "title": "The Hitchhiker's guide to the galaxy",
        },
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 50,
          "title": "The Hitchhiker's guide to the galaxy 2",
        },
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 42,
          "pages": Array [
            Object {
              "content": "blabla",
            },
          ],
          "title": "The Hitchhiker's guide to the galaxy",
        },
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 50,
          "pages": Array [
            Object {
              "content": "blabla",
            },
          ],
          "title": "The Hitchhiker's guide to the galaxy 2",
        },
      },
    },
  ],
}
`;

exports[`encodeParamsInResource tests encodeParamsInResource test 1 1`] = `"books_._._._{\\"fields\\":[\\"id\\",\\"name\\",\\"year\\"],\\"join\\":[{\\"field\\":\\"pages\\",\\"select\\":[\\"number\\",\\"words\\"]},{\\"field\\":\\"author\\",\\"select\\":[\\"id\\",\\"name\\"]},{\\"field\\":\\"author.favoriteFood\\",\\"select\\":[\\"id\\",\\"name\\"]}]}"`;

exports[`encoded params tests encoded params test 1 1`] = `
Object {
  "data": Object {
    "id": 1,
  },
}
`;

exports[`encoded params tests encoded params test 1 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://test.test/api/books/5?join[]=pages||number,words&join[]=author||id,name&join[]=author.favoriteFood||id,name&fields=id,name,year",
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "json": Object {
          "id": 1,
        },
      },
      "wasPromise": true,
    },
  ],
}
`;
